### DISCUSSION TITLE
how to use libmseed

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#179] [Design & Development]
I had another look at libmseed with a fresh eye looking at the problem.   I think we can make this work.   I have this plan:
1.  First, I'm going to hack the examples (something I apparently hadn't checked carefully  the last time I looked at this or I wouldn't have been as bewildered) and build a program that will output a set of lines with the following content:

   net sta chan loc foff length

2.  If I can make that work I'll turn it into a C++ function that returns a vector of struct values with the above attributes.   
3. with that I can build pybind11 wrappers and build a file indexing function.   That function could be parallelized as well with an input RDD/bag of a list of file names to be indexed.   

What do you think?

**Comment by wangyinz** [comment]
sounds good to me!

**Comment by pavlis** [comment]
This was much easier than I thought it would be once I got into the weeds.   I have items 1 and 2 done, but could use some input before I proceed to building the bindings to python with pybind11.  
1.  Where should this go?   Maybe we need a new ccore module called "import" as I can see good reasons for similar needs for other data formats.   Right now the module would only have this one function.
2. Building this into mspass is a challenge because of a decision point about how to do it.  For the prototype I just build the C (proto 1) and C++ (final for part 2) in a directory under libmseed that used the same build structure as the "examples" directory.   The complexity is that I am positive obspy already has a compiled libmseed (I can see it in their lib directory) but what is there appears to have python bindings linked into the .so file.   It is a bit of adventure land to try to link a C++/pybind11 set of code against a foreign library like that.   If we don't do that, however, the cmake setup will need to be modified to pull and install libmseed.   I think if we do that and only link against a static library we won't create a storage bloat problem.  This function uses only a tiny fraction of libmseed library (well actually I can't tell how interconnected it is but it isn't a huge library anyway).  

In any case I'm going to need some help in constructing this.   Step 1 is to decide where this should go in the mspass tree.   I can then  fork a branch from master and copy the current C++ function there. I can push the branch and maybe work this out collectively.

**Comment by wangyinz** [comment]
Wow, that was quick! I was looking at how ObsPy organizes the libmseed related files, and found something that seems quite relevant. The first thing I noticed is these lines that indicates the deficiency in reading large data that we observed is a known issue: 
https://github.com/obspy/obspy/blob/492b1e1709ce51faf023901be0f133677fdac2a1/obspy/io/mseed/core.py#L272-L276

Then, I followed the link there and found this example (https://github.com/obspy/obspy/pull/1419/#issuecomment-221582369) on how to read miniseed files in chunks using the ObsPy's API. This seems to be a much easiler route to proceed. We just need to a way to determine the chunk size to read in our API. What do you think?

**Reply by pavlis** [reply]
Interesting example.   On the other hand, the custom code I just developed has another very important advantage.   I can allow reading raw miniseed data in parallel and may, in fact, allow us to use the raw miniseed data as an origin a workflow.  I think we may now want to consider a wf_miniseed collection that can overlap strongly with wf_TimeSeries.   Maybe it is time to start the design document for how we can do that?

**Comment by wangyinz** [comment]
Yep, we can certainly do that. Let's first resolve how to get libmseed into our build structure. Maybe the easiest way is for me to setup the necessary CMake configs. I am not sure where should we put the prototype code at the moment, so maybe you can just email me the source code and I can incorporate them into a new branch for us to work on.

**Comment by wangyinz** [comment]
Well, actually, I have added libmseed to our cmake setup in the [add_libmseed](https://github.com/wangyinz/mspass/tree/add_libmseed) branch. You should now be able to use the `MSEED_LIBRARIES` and `MSEED_INCLUDE_DIR` to properly build code with libmseed dependency (totally untested though).

**Comment by pavlis** [comment]
Allright, I'm checking this out.   For starters I think I'll create an import libray.   Then the C++ low level stuff would be under mspasspy.ccore.import.  The reader I have in mind will have a python front end to mesh it more cleanly into mspass.  Here is a tentative signature:
```
def import_mseed(db,mseedfile,collection="wf_miniseed"):
```
where db is a Database as usual, mseedfile is the file to be indexed, and collection is where the index data is to be written.   Details of the schema for wf_miniseed need to be defined.  We do need to probably start a design document for that, but let me suggest how it would be set up.
1.  the start of wf_miniseed schema is equivalent to what in C would be called an "include" of the wf_TimeSeries schema..  i..e wf_miniseed is wf_TimeSeries with some additional attributes.   I think yaml has this capability directly with this syntax `<<: *wf_TimeSeries` right?   
2. We would need to add net, sta, and chan as required attributes and have loc be optional to mesh with standard usage.  
3. We would also need to require the following:   dir, dfile, foff, and nbytes.  The first there are as in css3.0.  The "nbytes" attribute is a little weird because with compression and the formatting of miniseed you need the number of bytes to read.  That size is not predictable from the number of samples or the time span.   The python code would need to use the methods for reading raw binary bytes of that size - equivalent to fread in C/C++ (I don't remember the python io function to do that but I know it exists and is fast).  We could then call the obspy mseed reader pointing at the internal buffer instead of a file - I know it supports that.  
4. I would add an oddity I called "last_pack_time" in the C++ code I wrote.  the oddity exists because the last_packet_time is almost and endtime but not exactly.   The anomaly is needed because miniseed files are nothing more than fixed length packets with a header BUT the data are usually compressed.  That makes it impossible to determine endtime exactly without decompressing the packet.  I could do that in the C code, but I thought it was unnecessary baggage since endtimes in mspass are computed with a method and not set as an attribute.   As an approximate end time the quantity I did save should work for range queries.  The only way it would be awkward is ultra low sample rate data like Ross Caton used in his PhD (0.1 sps) where the packets can span hours.  Those data volumes are so tiny, however, that the need for indexing is unlikely to be an problem.

**Comment by pavlis** [comment]
I just discovered an interesting disconnect.  There is a wide variance between the "latest release" (2.19.6) with a date tagged at 2018 and the version actually visible for browsing and downloading through github - it shows version 3.0.8.   I built my program linking against the later (3.0.8) and it will not build under 2.19.6.  the reasons are obvious when you compare them, which you don't need to do as I can tell you they definitely are very different.  I think all the changes since the latest release are related to url based io.   

Before I bug Chad about this, is there a way to have cmake run git clone on the repository instead of fetching the release tar ball.  For reference, this is what you have now in the cmake directory in the file called mseed-download.cmake:
```
cmake_minimum_required(VERSION 3.4)
  
project(mseed-download NONE)

include(ExternalProject)

ExternalProject_Add(
  mseed
  SOURCE_DIR "@MSEED_DOWNLOAD_ROOT@/mseed-src"
  URL
    https://github.com/iris-edu/libmseed/archive/refs/tags/v2.19.6.tar.gz
  URL_HASH
    SHA256=80b7f653589a30dcb4abcd53fe6bc1276f634262ee673815e12abdc1ff179d1d
  CONFIGURE_COMMAND ""
  BUILD_COMMAND make -j 8
  BUILD_IN_SOURCE 1
  INSTALL_COMMAND cp libmseed.a libmseed.h ${PROJECT_BINARY_DIR}
  TEST_COMMAND ""
  )
```
i.e. is there an incantation to have cmake run git clone instead of just a wget or whatever it does?

**Reply by wangyinz** [reply]
Yeah, I wasn't sure whether you need the latest version or its latest release when setting that up. Anyway, I just pushed the changes and now it will build the latest pre-release version, which is also the latest version of the master branch.

**Reply by pavlis** [reply]
Ok, thanks for getting to that so quickly.   I'll be curious to look how you did that.   Makes sense that that functionality should be there.

**Comment by pavlis** [comment]
Well, I have this almost finished.  The C++ code compiles and I have a pybind11 file that compiles.  I'm having linking problems, however, and wonder if you, Ian, can reproduce the problem.   Here is what I'm getting linking the current add_libmseed branch:
```
[ 78%] Linking CXX shared module importers.cpython-38-x86_64-linux-gnu.so
/usr/bin/ld: ../../mseed/libmseed.a(fileutils.o): relocation R_X86_64_PC32 against symbol `gMS3FileParam' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [python/importers/CMakeFiles/pyimport.dir/build.make:102: python/importers/importers.cpython-38-x86_64-linux-gnu.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:728: python/importers/CMakeFiles/pyimport.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```
As the message hints I think this would be fixed if we could compile libmseed with -fPIC but I can't quite figure out how you would do that.  I suspect it is some change  to the cmake/mseed.cmake file, but that is a guess.  

If I can get past this linking problem I thimk I can write function to use it to build an index pretty quickly.   Hoping you can help.

**Reply by wangyinz** [reply]
I got that fixed in https://github.com/wangyinz/mspass/commit/eedf778e16e28b3eb8ad2769a7d663e8704fa463. Also, I noticed the collision of keyword `import` under Python, so I change the name of the new module to `io`.

**Reply by pavlis** [reply]
Glad you did that.  I had run into the import keyword collsion earlier.  I like io a lot better than importer.  I'll give this a go.   Thanks for taking care of this so quickly.

**Reply by pavlis** [reply]
Hmm - I need to punt to you again.   I might be able to figure this out, but am hoping you will know how to do this immediately.  If not feel free to punt it back to me.

Problem is the code for _mseed_file_indexer is not being linked into the no io module.  Weird because the pybind11 code requires it, but dir on the io module show it is not defined an import confirms it.  I must be missing something but I assumed the Cmake file for the io pybind11 directory should be a modification of the comparable modules (i.e. seismic, utility, and algorithms).  I must be missing something obvious.

BTW - your fix for libmseed worked fine, BUT I had to rebuild the whole tree.   Is there a way to avoid that if the configuration for a download package changes?  cmake doesn't seem to be able to detect such changes and the only way I've found in the past, as here, to resolve such problems it to `rm -r build; mkdir build; cd build; ccmake ..`

**Reply by wangyinz** [reply]
I just tested on my machine, and I was able to import and access that function fine:

```
>>> import mspasspy.ccore.io
>>> mspasspy.ccore.io._mseed_file_indexer()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: _mseed_file_indexer(): incompatible function arguments. The following argument types are supported:
    1. (file: str) -> mspasspy.ccore.io.MseedIndex

Invoked with: 
>>> 
```

Yes, you need to rebuild the libmseed. Sorry that I forgot about that issue. Actually, you can just delete the `build/mseed` directory, and CMake will re-download and rebuild mseed correctly.

**Reply by wangyinz** [reply]
I also checked our libmspass and can see the symbols are correctly defined: 
```
$ nm src/lib/libmspass.a | grep mseed_file
mseed_file_indexer.o:
0000000000000aab t _GLOBAL__sub_D_00100_1_mseed_file_indexer.cc
0000000000000a99 t _GLOBAL__sub_I_00100_0_mseed_file_indexer.cc
0000000000000a60 t _GLOBAL__sub_I_mseed_file_indexer.cc
0000000000000000 T _ZN6mspass2io18mseed_file_indexerENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
0000000000000010 b __gcov0._GLOBAL__sub_I_mseed_file_indexer.cc
0000000000000d80 b __gcov0._ZN6mspass2io18mseed_file_indexerENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
0000000000000000 d __gcov_._GLOBAL__sub_I_mseed_file_indexer.cc
0000000000001a80 d __gcov_._ZN6mspass2io18mseed_file_indexerENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```
Is there something that I missed?

**Reply by pavlis** [reply]
I made a dumb mistake here.   I had only build the io module using `make` in the cxx/build directory.  Then I did a cd to the io directory and tried:  `import io`.   A more expert python programmer would have realized the symbols I got when I did a dir(io) were those of the python io module.   Apparently that took precedence over the io.cython.blah.so file in the current directory.   That actually totally makes sense when I check the docs as io is a standard python library.  

Maybe we need a different name yet?  To allow for future expansion maybe we should use  "import_export"?   Don't think that will create a name collision.   We can decide that later.  I can now test this and move to  building a reader.

**Comment by pavlis** [comment]
Slap me up on this one.   I was caught in another symbol collisions.  I had just run make in the build and did a cd to the io directory.  Then tried “import io” BUT (this is where I deserve to be slapped) that got me the standard python “io” module NOT what I really needed which was (once I ran setup.py):
import mspasspy.ccore.io<http://mspasspy.ccore.io>
that works.

Sorry for that dumb mistake.  I’ll leave a record of this in discussion

On Apr 19, 2021, at 8:50 AM, Ian Wang ***@***.******@***.***>> wrote:

This message was sent from a non-IU address. Please exercise caution when clicking links or opening attachments from external sources.



I also checked our libmspass and can see the symbols are correctly defined:

$ nm src/lib/libmspass.a | grep mseed_file
mseed_file_indexer.o:
0000000000000aab t _GLOBAL__sub_D_00100_1_mseed_file_indexer.cc
0000000000000a99 t _GLOBAL__sub_I_00100_0_mseed_file_indexer.cc
0000000000000a60 t _GLOBAL__sub_I_mseed_file_indexer.cc
0000000000000000 T _ZN6mspass2io18mseed_file_indexerENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
0000000000000010 b __gcov0._GLOBAL__sub_I_mseed_file_indexer.cc
0000000000000d80 b __gcov0._ZN6mspass2io18mseed_file_indexerENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
0000000000000000 d __gcov_._GLOBAL__sub_I_mseed_file_indexer.cc
0000000000001a80 d __gcov_._ZN6mspass2io18mseed_file_indexerENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE


Is there something that I missed?

**Comment by pavlis** [comment]
I have this working.   Now a couple followup questions:
1.   to produce a proper test script I need to have a stock mseed file.  the one I have is  much larger than necessary.  Easy to produce a smaller test file, but the real question is where the data file should go and how the test should be set up for that kind of thing?  Actually, come to think of it we could probably have the test do an obspy download and run the test on the downloaded file?   Please advise how you want to proceed on that.
2. We may want to discuss the schema for an mseed reader, but propose I simply build one for testing of the reader I'm working on and we can use it as the "strawman" to finalize a design.   Probably more efficient that way than commenting on a document out of context.  Let me know if you have another strategy.
3. I'm going to use obspy's reader in with the file like interface.  It should work, but there will be the overhead of then converting the Trace object to TimeSeries.

**Reply by wangyinz** [reply]
1. Download on the fly is never a good idea for testing. We should prepare a small miniseed file and put it under `python/tests/data`. There are already a bunch of other files for testing there.
2. Sounds good!
3. Yeah, we will see whether the overhead is significant and decide if having our own reader in cxx is perferred.

**Comment by pavlis** [comment]
Ok, I just pushed a draft to the add_miniseed branch.  I'll issue a pull request after I verify it install properly with pip.   A few points and holes:
1.  I am not confident I modified the yaml file correctly for the new collection I've called wf_miniseed.  Not even sure how to test if it is ok.
2.  I put the python prototypes in mspasspy.io with the name mseed.py.   Feel free to change the name if you want.  
3.  I don't know how to build a test for these two functions.  I had a basic miniseed file and first ran the following little function to verify the indexer was working properly:
```
import mseed
from mspasspy.db.client import DBClient 
from mspasspy.db.database import Database 
dbclient=DBClient()
db=Database(dbclient,'getting_started')
dfile='test.mseed'
mseed.index_mseed_file(db, dfile)
```
then I ran this one to verify the reader:
```
from mspasspy.io.mseed import dbread_miniseed
from mspasspy.db.client import DBClient 
from mspasspy.db.database import Database 

dbclient=DBClient()
db=Database(dbclient,'getting_started')
curs=db.wf_miniseed.find({}).limit(10)
count=0
for doc in curs:
    print('This is the doc content for doc number ',count)
    print(doc)
    ts=dbread_miniseed(db,doc,load_history=True)
    print('time series str representation of data read:')
    print(ts)
```
These ran on my local machine, but they are not suitable for a test suite.   Can I punt to you building a test with a simple mseed file of your choosing?

I have a residual issue we might want to defer to our next weekly meeting.  I stubbed in code for a "load_history" option for the reader but didn't know how to set the algid.

Anything else - see the pull request.

**Reply by wangyinz** [reply]
Sure, I can take care of the tests. In the meantime, I think we probably should move those new functions to the db module or change their behavior. This is because usually the name `io` refers to reading/writing files, but the new functions are actually doing database operations mostly. At the moment, it is probably more straightforward to just move them, but we might want to implement a miniseed reader that takes the input of `foff` and `nbytes`. In that case, we probably will place it under the `ccore.io` module anyway.

Another question I have is the following error message when calling the `_mseed_file_indexer` function:
```
source id string= could not be decoded - skipping one or more packets
```
I tried it with multiple miniseed files that I downloaded using the FDSN web service, and it always appears once no matter how many channels the file has. I am not sure what this error complains but it does look like a bug to me. You should be able to reproduce the error using the sample data that I just added to the repo at `python/tests/data/3channels.mseed`.

**Reply by pavlis** [reply]
1.  I wondered if it putting these in database.py was needed.  Will do.  Does raise a question though I don't fully understand in python.   That is, are there consequences to class bloat?  As I understand it in a compiled OOP language like C++ there isn't really any because a method is just a special function in terms of the linked executable.  Further, for modern machines the memory size is so huge it has become the norm to never worry about the size of the code segment.    It was a lot different in the days of 64 K memory, so I still worry about memory use.   Bear with me - do you know if there are efficiency or memory tradeoffs in python in procedural functions like those I produced here versus making the same function part of a class? 
2. I had noticed that error too, but had (apparently incorrectly) assumed it was an anomaly in the file I was using for testing.  It apparently is not.  I had tried to run a tedious antelope program called msdd that dumps the header of every packet to try to find the bad block, but there was too much information to find the needle in a haystack (i.e one block in a million).   I'll see if I can figure this out.   I have a suspicion it is a different seed blockette that is not the special packet type that defines a data block.  If so, I will need to ping dmc people for an explanation.  For now let me see if I can figure out what is doing that.

**Reply by pavlis** [reply]
Well, I figured the bad sid error out quickly.   There is a bug in handling the first block of data.   Why exactly is yet to be determined but the fundamental reason is the really weird way libmseed handles end of file and the way the examples say to construct a read loop.  I don't expect it to be a hard problem to solve.  Hopefully I'll have all this finished by the time you get up - you were up really late late night from the time stamp.

**Reply by pavlis** [reply]
Done and bug in first block was resolved.  Pushed the changes. 

Do note [this issue](https://github.com/wangyinz/mspass/issues/181) I posted that I discovered adding this to Database.  Advise you address that in the  add_libmseed branch before merging it to master.

**Reply by wangyinz** [reply]
Well, I am not sure about the tradeoff issue in Python, but to my understanding it shouldn't be a problem making them methods of the database class.

Glad you resolved the error quickly! I checked the issue about database you posted and do think that is a bug. I think that problematic line was a copy-and-paste from the `update_metadata` method below. @haruming should be able to fix it quickly.

**Comment by wangyinz** [comment]
I took a closer look at the new `read_miniseed` method, and I think we should probably merge that into the `read_data` method and define something similar to the `_read_data_from_dfile` method that reads a miniseed trace. If so, I guess we also need to add an optional keyword to the wf collection to specify the file format. What do you think?

**Comment by wangyinz** [comment]
I went ahead and implemented this idea in https://github.com/wangyinz/mspass/pull/180/commits/13a35014c1d2700386109268d31c895fc435f881. However, I run into a problem which is actually a conflict in our reader design. Basically, the `wf_miniseed` documents created by the `index_mseed_file` function do not include any of the core metadata (i.e., `dt`, `t0`) needed to construct our data objects. Therefore, calling `read_data` over the `wf_miniseed` collection will cause an error in the `TimeSeries` constructor. To properly address this issue, I guess we will have to also pull those required metadata from the miniseed files in the `_mseed_file_indexer` function, and then keep them in the database. Any other thoughts?

**Reply by pavlis** [reply]
Not sure that is an error in the design of read_data.   I think this illustrates that we should cleanly separate data import methods from the core wf processing collections.   We had some lengthy discussion on this early on in the development and we are looping back.   You know more about finding things like that in our earlier discussions but on this one my memory is not faulty.   I think it important to cleanly separate import functions from processing functions as they are conceptually different.   That is one key difference - some formats have concepts that do may not match a given schema.  For example, we haven't even attempted to consider active source data that are imported as some segy variant.   

What this does suggest is maybe we should rename the miniseed read function `import_mseed`?  

Another idea I considered before I reverted to a procedural approach to build the prototypes for tha stuff is to use inheritance.   That is, maybe it would be cleaner to create a `MSEEDDatabase` that is a subclass/child of our Database class?   It's default schema would include wf_miniseed and would add the methods in the new add_miniseed branch.  A key reason I didn't do it was I got confused about how inheritance works in detail in python.   I may have been reading the wrong sources but I got confused so decided it was better to get this done.  Further, at the time I thought I shouldn't just do that without discussing it as it is an important design decision.

The key advantage of an inheritance mechanism is it would provide a clean solution to import other formats and not worry about name collisions.   A secondary issue I didn't understand in python that made me punt this idea earlier is how python handles overriding methods.   i.e. one might want to override the read_data method in the subclass to support miniseed automatically without having to touch the base class Database.   Do either of you know how to do that?  

Seems this is likely a good topic for discussion in today's meeting.
